### PR TITLE
Feature/486 bri-3 workflow crud api and tests

### PR DESCRIPTION
**Pull request by Kasshern**
# Description

This PR introduces CRUD API for workflows, creates a many to many relation between workflow and workstep, and introduces necessary tests: Completing #486 

## Related Issue

#486 

## Motivation and Context

Introduces basic CRUD for workflow type of BRI entity

## How Has This Been Tested

Added unit tests. Manual testing through submitted postman collection

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+          w.workgroupId,
+          w.securityPolicy,
+          w.privacyPolicy,
+        );
+      }),
+      workflowModel.workgroupId,
+    );
+  }
+
+  async getAllWorkflows(): Promise<Workflow[]> {
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:37]: again scope? @Kasshern

File: examples/bri-3/src/bri/workgroup/workflows/models/workflow.ts [hunk 2]
```diff
@@ -1,11 +1,48 @@
 import { Workstep } from '../../worksteps/models/workstep';
 
 export class Workflow {
-  private id: string; // TODO: Add uuid after #491
-  private worksteps: Workstep[]; //TODO enforce workstep causal connection through collection order
+  private _id: string; // TODO: Add uuid after #491
+  private _name: string;
+  private _worksteps: Workstep[]; //TODO enforce workstep causal connection through collection order
```
Comment (Therecanbeonlyone1969) [inline examples/bri-3/src/bri/workgroup/workflows/models/workflow.ts:6]: @Kasshern might be easiest to enforce order through a binary sparse merkle tree of worksteps organized into a workflow. just a thought.

**Review by Therecanbeonlyone1969 — CHANGES_REQUESTED** [review]
Looks really good. great work. see my minor comments. @Kasshern

File: examples/bri-3/prisma/migrations/20220930160450_add_workflow_table/migration.sql [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+CREATE TABLE "Workflow" (
+    "id" TEXT NOT NULL,
+    "name" TEXT NOT NULL,
+    "workgroupId" TEXT NOT NULL,
+
+    CONSTRAINT "Workflow_pkey" PRIMARY KEY ("id")
+);
+
+-- CreateTable
+CREATE TABLE "_WorkflowToWorkstep" (
```
[the remark this answers is not in the document]
Comment (Kasshern) [reply examples/bri-3/prisma/migrations/20220930160450_add_workflow_table/migration.sql:11]: The team has decided to stick with prisma's out-of-box implicit rules. Thus, the migration file's use is abstract and interfaced with through prisma commands.

File: examples/bri-3/src/bri/workgroup/workflows/models/workflow.ts (hunk 2 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/models/workflow.ts:6]: @Therecanbeonlyone1969 Will create a separate issue corresponding to this TODO and add your advice.

File: examples/bri-3/src/bri/workgroup/workflows/agents/mockWorkflowStorage.agent.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  async getWorkflowById(id: string): Promise<Workflow> {
+    const workflow = this.workflowsStore.find((ws) => ws.id === id);
+    if (!workflow) {
+      throw new NotFoundException(WORKFLOW_NOT_FOUND_ERR_MESSAGE);
+    }
+    return workflow;
+  }
+
+  async getAllWorkflows(): Promise<Workflow[]> {
```
[the remark this answers is not in the document]
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/mockWorkflowStorage.agent.ts:18]: This mock agent is only for testing purposes will update it to reflect the query scope of the true agent once that is decided.

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts (hunk 1 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:37]: @Therecanbeonlyone1969 any suggestions what to query on? Retrieve those belonging to specific workgroup? or a specific name? including certain worksteps. A combination of those?

Comment (ognjenkurtic) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:37]: I am fine if we keep it as is for now. We 'll scope it as we move forward and get a better understanding of the needs of the BPI consumers.

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      data: {
+        id: workflow.id,
+        name: workflow.name,
+        worksteps: {
+          connect: workstepsData,
+        },
+        workgroupId: workflow.workgroupId,
+      },
+      include: {
+        worksteps: true,
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:78]: This means it is creating worksteps at the same time?

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+            ws.privacyPolicy,
+          );
+        }),
+        w.workgroupId,
+      );
+    });
+  }
+
+  async createNewWorkflow(workflow: Workflow): Promise<Workflow> {
+    const workstepsData = workflow.worksteps.map((w) => {
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:62]: why is this needed? variable name is a little bit too general to understand the reasoning behind this code.

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  }
+
+  async updateWorkflow(workflow: Workflow): Promise<Workflow> {
+    const workstepsData = workflow.worksteps.map((w) => {
+      return {
+        id: w.id,
+      };
+    });
+
+    const newWorkflowModel = await this.workflow.update({
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:107]: updatedWorkflowModel or same naming like in other update methods

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      where: { id: workflow.id },
+      data: {
+        name: workflow.name,
+        worksteps: {
+          set: workstepsData,
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:112]: why do we set set to workstep ids?

File: examples/bri-3/src/bri/workgroup/workflows/api/workflows.controller.spec.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+      // Act and assert
+      expect(async () => {
+        await workflowController.updateWorkflow(nonExistentId, requestDto);
+      }).rejects.toThrow(new NotFoundException(WORKFLOW_NOT_FOUND_ERR_MESSAGE));
+    });
+
+    it('should perform the update if existing id passed', async () => {
+      // Arrange
+      // const workstepRequestDto = {
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workflows/api/workflows.controller.spec.ts:184]: can be deleted?

**Review by ognjenkurtic — DISMISSED** [review]

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflows.agent.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  WORKSTEP_NOT_FOUND_ERR_MESSAGE,
+} from '../api/err.messages';
+import { WorkflowStorageAgent } from './workflowsStorage.agent';
+import { WorkstepStorageAgent } from '../../worksteps/agents/workstepsStorage.agent';
+
+@Injectable()
+export class WorkflowAgent {
+  constructor(
+    private workflowStorageAgent: WorkflowStorageAgent,
+    private WorkstepStorageAgent: WorkstepStorageAgent,
```
[the remark this answers is not in the document]
Comment (Manik-Jain) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflows.agent.ts:16]: +1

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts (hunk 5 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:78]: This is just including the worksteps connected in the immediate above lines in the returned object of the query

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts (hunk 1 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:37]: Will leave as is and scope the query when BPI consumer needs elucidated.

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflows.agent.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflows.agent.ts]: fixed

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts (hunk 7 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts]: fixed

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts (hunk 6 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts]: rename the workstepsData to workstepIds. This is pulling out  just the ids of the worksteps within workflow because that is all prisma needs to add these relationships to the join table within the DB that reflects the many to many relationship between worksteps and workflow

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts (hunk 8 above)
Comment (Kasshern) [reply examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts]: Just like "connect" in the createNewWorkflow function above prisma only needs the workstepIds to either create or update the join table of the many-to-many relationship between worksteps and workflows. I have renamed workstepsData to workstepIds here as well. Set will update those entries in the join table of the DB.

**Review by ognjenkurtic — APPROVED** [review]

**Review by Therecanbeonlyone1969 — APPROVED** [review]
LGTM now
